### ISSUE TITLE
Same RequestMapping URL's containing different RequestParam gets overridden 

### ISSUE THREAD

**Issue by rajeshkamal** [springfox/springfox#750]
Hi,

The below user controller contains two GET methods mapped to the same URL. One of the method contains mapping to a specific request param "status"

```
@RestController
@Api(value = "USER")
@RequestMapping("/users")
public class UserController {
    
    @RequestMapping(method = RequestMethod.GET, produces="application/json")
    @ResponseStatus(value= HttpStatus.OK)
    public Collection<User> getUsers() {
        throw new NotImplementedException();
    }

    @RequestMapping(method = RequestMethod.GET, produces="application/json", params={"status"})
    @ResponseStatus(value= HttpStatus.OK)
    public Collection<User> getUsersByStatus(@RequestParam(value="status") String status) {
        throw new NotImplementedException();
    }

}
```
Couple of issues,
1. The generated spec contains only 1 misses the other. Similarly if I have more methods to the same URL with different query params/request params only 1 is available. All of the variations should be made available.
2. Here even the generated 1 operation/method's spec is wrong. The method contains only 1 status query param. But the spec contains 2 status query param. This is possibly because of the @RequestMapping(params={"status"}) and @RequestParam(value = "status") String status. But they are one and the same. It is done like this because to narrow down the call.

```
    "/api/users": {
      "get": {
        "tags": [
          "user"
        ],
        "summary": "Get the list of users by status",
        "description": "getUsersByStatus",
        "operationId": "getUsers",
        "consumes": [
          "application/json"
        ],
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "name": "status",
            "in": "query",
            "required": true,
            "type": "string"
          },
          {
            "name": "status",
            "in": "query",
            "description": "status",
            "required": true,
            "type": "string"
          }
        ],
        "responses": {
          "200": {
            "schema": {
              "$ref": "#/definitions/Resources"
            }
          },
          "500": {
            "description": "Internal Server Error",
            "schema": {
              "$ref": "#/definitions/Error"
            }
          }
        }
      }
    }
```
@dilipkrish  @adrianbk 
Regards,
Rajesh Kamal

**Comment by dilipkrish** [comment]
Thanks again for catching an esoteric bug :)

**Comment by rajeshkamal** [comment]
Thanks !! This seems to be a very critical issue as some of the API's are not made available at the spec level. Which affects the Swagger UI, Code generation etc. It will be great if you can get this fixed in the upcoming milestone.

**Comment by dilipkrish** [comment]
This really depends on how this issue https://github.com/swagger-api/swagger-spec/issues/291 turns out. Might provide support for this in a naive way before that happens, however I doubt that swagger-ui will support it.

**Comment by rajeshkamal** [comment]
Do you really think it is related to the above issue. Can't we get the spec to be generated like the below.

```
"/api/users": {
      "get": {
        "tags": [
          "user"
        ],
        "summary": "Get the list of users",
        "description": "getUsers",
        "operationId": "getUsers",
        "consumes": [
          "application/json"
        ],
        "produces": [
          "application/json"
        ],
        "responses": {
          "200": {
            "schema": {
              "$ref": "#/definitions/Resources"
            }
          },
          "500": {
            "description": "Internal Server Error",
            "schema": {
              "$ref": "#/definitions/Error"
            }
          }
        }
      }
    },
"/api/users": {
      "get": {
        "tags": [
          "user"
        ],
        "summary": "Get the list of users by status",
        "description": "getUsersByStatus",
        "operationId": "getUsersByStatus",
        "consumes": [
          "application/json"
        ],
        "produces": [
          "application/json"
        ],
        "parameters": [
          {
            "name": "status",
            "in": "query",
            "description": "status",
            "required": true,
            "type": "string"
          }
        ],
        "responses": {
          "200": {
            "schema": {
              "$ref": "#/definitions/Resources"
            }
          },
          "500": {
            "description": "Internal Server Error",
            "schema": {
              "$ref": "#/definitions/Error"
            }
          }
        }
      }
    }
```

**Comment by dilipkrish** [comment]
Unfortunately swagger spec does not support non-unique paths. If you look at the datastructures used by swagger models, you'll see that operations are represented by using a map going from path (string) to `Path`  (operation).

**Comment by GitVhaos** [comment]
+1 from my side! This is a really important feature! Otherwise no "search" operations are supported with swagger.

**Comment by rajeshkamal** [comment]
@dilipkrish @adrianbk 

Do you have any workaround for this. I am currently blocked on this.

Because the generated spec and in turn the generate java code generated does not even compile.

...
public List<User> getUsersByStatus (String status, String status) throws ApiException {
...

Kindly appreciate your inputs or workarounds on this.

**Comment by dilipkrish** [comment]
I know it will be missing some operations but it's not clear why it won't compile. 😕

**Comment by rajeshkamal** [comment]
The generated spec and in turn code contains duplicate params.

RestController
```
    @RequestMapping(method = RequestMethod.GET, params = {"status"})
    public Collection<User> getUsersByStatus(
            @RequestParam(value="status") String status) {
    }
```

Swagger Spec

```
        "parameters": [
          {
            "name": "status",
            "in": "query",
            "required": true,
            "type": "string"
          },
          {
            "name": "status",
            "in": "query",
            "description": "status",
            "required": true,
            "type": "string"
          }
        ],
```

Generated Code - DOES NOT COMPILE
```
public List<User> getUsersByStatus (String status, String status)
```

**Comment by dilipkrish** [comment]
Could you create a new issue for this? It has nothing to do with multiple request mapping a with  duplicate paths

**Comment by dilipkrish** [comment]
In the mean while to get u up and running it should work if you remove the `params`  attribute

**Comment by rajeshkamal** [comment]
Thanks, I will create a separate bug for this issue.

The controllers come to us as part of a different component which we do not own. Also it is not in 1 place, and it might be difficult to get it fixed across teams with different components.

It will be great if you can get this https://github.com/springfox/springfox/issues/813 fixed

**Comment by dilipkrish** [comment]
This is now enabled via incubating feature to enableUrlTemplating via the Docket. Please provide feedback if you're using it.

**Comment by rajeshkamal** [comment]
I just tried the enableUrlTemplating feature and it is really nice @dilipkrish 

But, I am bumping into these issues,
1. #1097 Swagger UI. I have updated the bug there.
2. Potentially might bump into swagger code generation issues with url templating.  Any inputs on how I can overcome/workarounds for this one.

**Comment by dilipkrish** [comment]
@rajeshkamal regarding 1> you can try [another experimental web jar](http://mvnrepository.com/artifact/io.springfox/springfox-swagger-ui-rfc6570/0.9.4) with rfc6570 support. 
**Note:** _Keep in mind this is experimental_!

Regarding 2> You're on your own unfortunately. Considering these are simple wrappers around _curl_ operations, its not worth the effort supporting that. Plus this is just an incubating feature, not even officially part of the spec.

**Comment by orbfish** [comment]
It's very nice for one of my endpoints, but not great for the others, because it adds a bunch of information that is unnecessary when the multiple endpoints per uri/verb is not present. Would be great to be able to limit this per endpoint/resource class, as well has to have swagger ui respond correctly.

**Comment by dilipkrish** [comment]
@orbfish Please don't cross post on old issues.

**Comment by orbfish** [comment]
I'm sorry - a search brought up a number of related issues, and I wasn't sure where was the best place to reply.

**Comment by dilipkrish** [comment]
@orbfish not a problem. If you see that an issue is closed. Its always better to create a new issue referencing the issues you believe the new issue is related to. Plus the added advantage is you'll be notified if *your* issue is fixed.
